src/Apn.php

```diff
@@ -6,167 +6,284 @@
     {
-        if (isset($this->config['connection_attempts']) &&
-            is_numeric($this->config['connection_attempts'])) {
-            $this->maxAttempts = $this->config['connection_attempts'];
-        }
+        return self::APNS_PRODUCTION_SERVER . $this->getUrlPath($deviceToken);
     }
 
     /**
-     * Determines whether the connection attempts should be unlimited.
+     * Get Url for APNs sandbox server.
      *
-     * @return bool
+     * @param Notification $notification
+     * @return string
      */
-    private function isUnlimitedAttempts()
+    private function getSandboxUrl(string $deviceToken)
     {
-        return $this->maxAttempts == 0;
+        return self::APNS_DEVELOPMENT_SERVER . $this->getUrlPath($deviceToken);
     }
 
     /**
-     * Check if can retry a connection
+     * Get Url path.
      *
-     * @return bool
+     * @param Notification $notification
+     * @return mixed
      */
-    private function canRetry()
+    private function getUrlPath(string $deviceToken)
     {
-        if ($this->isUnlimitedAttempts()) {
-            return true;
-        }
-
-        $this->attempts++;
-
-        return $this->attempts < $this->maxAttempts;
+        return str_replace("{token}", $deviceToken, self::APNS_PATH_SCHEMA);
     }
 
     /**
-     * Reset connection attempts
+     * Decorate headers
      *
-     * @return $this
+     * @return array
      */
-    private function resetAttempts()
+    public function decorateHeaders(array $headers): array
     {
-        $this->attempts = 0;
-
-        return $this;
+        $decoratedHeaders = [];
+        foreach ($headers as $name => $value) {
+            $decoratedHeaders[] = $name . ': ' . $value;
+        }
+        return $decoratedHeaders;
     }
 
     /**
-     * Provide the unregistered tokens of the notification sent.
-     *
-     * @param array $devices_token
-     * @return array
+     * @param $token
+     * @param array $message
+     * @param $request
+     * @param array $deviceTokens
      */
-    public function getUnregisteredDeviceTokens(array $devices_token)
+    public function prepareHandle($deviceToken, array $message)
     {
-        $tokens = [];
+        $uri = false === $this->config['dry_run'] ? $this->getProductionUrl($deviceToken) : $this->getSandboxUrl($deviceToken);
+        $headers = $message['headers'];
+        unset($message['headers']);
+        $body = json_encode($message);
+
+        $config = $this->config;
+
+        $options = [
+            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_2,
+            CURLOPT_URL => $uri,
+            CURLOPT_PORT => self::APNS_PORT,
+            CURLOPT_POST => true,
+            CURLOPT_POSTFIELDS => $body,
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_TIMEOUT => 10,
+            CURLOPT_HEADER => true,
+
+            CURLOPT_SSLCERT        => $config['certificate'],
+            CURLOPT_SSLCERTPASSWD  => $config['passPhrase'],
```

Comment:
potential undefined index

Reply:
How do you use it?

Reply:
{\"reason\":\"MissingTopic\"}

Reply:
```
        $message = [
            'headers' => [
                'apns-topic' => config('push.apnTopic'),
                'apns-push-type' => 'alert',
            ],
            'aps' => [
                'category' => $pushMessage->getKey(),
            ],
            'data' => $pushMessage->getData(),
        ];
```
like this

Reply:
I found that apnchannel does not have headers. I hope this can be added so that it can be sent through channel